bundles/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/internal/discovery/TeslaVehicleDiscoveryService.java

```diff
@@ -0,0 +1,110 @@
+import org.eclipse.smarthome.core.thing.binding.ThingHandlerService;
+import org.openhab.binding.tesla.internal.TeslaBindingConstants;
+import org.openhab.binding.tesla.internal.TeslaHandlerFactory;
+import org.openhab.binding.tesla.internal.handler.TeslaAccountHandler;
+import org.openhab.binding.tesla.internal.handler.VehicleListener;
+import org.openhab.binding.tesla.internal.protocol.Vehicle;
+import org.osgi.service.component.annotations.Component;
+
+/**
+ * This service is used by {@link TeslaAccountHandler} instances in order to
+ * automatically provide vehicle information from the account.
+ *
+ * @author Kai Kreuzer - Initial contribution
+ *
+ */
+@Component(service = ThingHandlerService.class)
+public class TeslaVehicleDiscoveryService extends AbstractDiscoveryService
+        implements DiscoveryService, VehicleListener, ThingHandlerService {
+
+    private static final Map<String, ThingTypeUID> modelMap = new HashMap<>();
+    static {
+        // see https://tesla-api.timdorr.com/vehicle/optioncodes
+        modelMap.put("MDLS", TeslaBindingConstants.THING_TYPE_MODELS);
+        modelMap.put("MS03", TeslaBindingConstants.THING_TYPE_MODELS);
+        modelMap.put("MS04", TeslaBindingConstants.THING_TYPE_MODELS);
+        modelMap.put("MDLX", TeslaBindingConstants.THING_TYPE_MODELX);
+        modelMap.put("MDL3", TeslaBindingConstants.THING_TYPE_MODEL3);
+        modelMap.put("MDLY", TeslaBindingConstants.THING_TYPE_MODELY);
+    }
+
+    public TeslaVehicleDiscoveryService() throws IllegalArgumentException {
+        super(TeslaHandlerFactory.SUPPORTED_THING_TYPES_UIDS, 10, true);
+    }
+
+    private TeslaAccountHandler handler;
+
+    @Override
+    public void setThingHandler(ThingHandler handler) {
+        this.handler = (TeslaAccountHandler) handler;
+        this.handler.addVehicleListener(this);
+    }
+
+    @Override
+    public ThingHandler getThingHandler() {
+        return handler;
+    }
+
+    @Override
+    protected void startScan() {
+        handler.scanForVehicles();
+    }
+
+    @Override
+    public void activate(@Nullable Map<@NonNull String, @Nullable Object> configProperties) {
+        super.activate(configProperties);
+    }
+
+    @Override
+    public void deactivate() {
+        super.deactivate();
+        if (handler != null) {
+            handler.removeVehicleListener(this);
+        }
+    }
+
+    @Override
+    public void vehicleFound(Vehicle vehicle) {
+        ThingTypeUID type = identifyModel(vehicle);
+        ThingUID thingUID = new ThingUID(type, handler.getThing().getUID(), vehicle.vin);
+        DiscoveryResult discoveryResult = DiscoveryResultBuilder.create(thingUID).withLabel(vehicle.display_name)
+                .withBridge(handler.getThing().getUID()).withProperty(TeslaBindingConstants.VIN, vehicle.vin).build();
+        thingDiscovered(discoveryResult);
+    }
+
+    private ThingTypeUID identifyModel(Vehicle vehicle) {
+        String options = vehicle.option_codes;
```

Comment:
So the tesla API is identifying my car with the following options
```
AD15,MDL3,PBSB,RENA,BT37,ID3W,RF3G,S3PB,DRLH,DV2W,W39B,APF0,COUS,BC3B,CH07,PC30,FC3P,FG31,GLFR,HL31,HM31,IL31,LTPB,MR31,FM3B,RS3H,SA3P,STCP,SC04,SU3C,T3CA,TW00,TM00,UT3P,WR00,AU3P,APH3,AF00,ZCST,MI00,CDM0
```
Which says I have a 75kwh Model 3, which is not correct.  According to this issue https://github.com/timdorr/tesla-api/issues/126#issuecomment-478613672, it may be that this is often not right (looks like a default model3 placeholder) and we should be querying the vehicle_config endpoint to get the car model.

If I hit the vehicle_config endpoint at `https://owner-api.teslamotors.com/api/1/vehicles/{{ids}}/data_request/vehicle_config
`  I get the following data
```json
{
    "response": {
        "can_accept_navigation_requests": true,
        "can_actuate_trunks": true,
        "car_special_type": "base",
        "car_type": "models",
        "charge_port_type": "US",
        "eu_vehicle": false,
        "exterior_color": "Black",
        "has_air_suspension": true,
        "has_ludicrous_mode": false,
        "motorized_charge_port": true,
        "plg": true,
        "rear_seat_heaters": 0,
        "rear_seat_type": 0,
        "rhd": false,
        "roof_color": "None",
        "seat_type": 1,
        "spoiler_type": "None",
        "sun_roof_installed": 1,
        "third_row_seats": "None",
        "timestamp": 1573106570396,
        "trim_badging": "p85d",
        "use_range_badging": false,
        "wheel_type": "Super21Gray"
    }
}
```
which correctly identifies my car as a p85D Model S.

Let me know if you want any help here, I can look at requesting this data instead of relying on the options in the `Vehicle` object , but I won't be able to get to it until tonight or tomorrow.

Reply:
Hm, that's interesting. Seems indeed that the other request is what we should do.
Feel free to do the change in the code through a PR against mine - I'll happily merge it.